mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java

```diff
@@ -101,7 +111,28 @@
 
     @Override
     public Publisher<Void> delete(User user, MessageId messageId) {
-        throw new NotImplementedException("Will be implemented later");
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
```

Comment:
How about:

```
metadatavault.retrieveStorageInformation(user, messageId)
   .flatMap(storageInformation -> metadatavault.remove(storageInformation.getBucket(), user, messageId)
       .then(Mono.just(storageInformation))
   .flatMap(storageInformation -> blobStore.delete(storageInformation.getBucket(), storageInformation.getMessageId())
```

 - 1. Retrieve the storage information
 - 2. Delete related metadata
 - 3. Delete the related blob

(Please note that we delete the reference to the blob first)

That way:
 - You don't need to scroll all buckets to find what you need
 - You rely on denormalization of the deleted message metadata vault

Reply:
fixed, sorry I have to rebase to resolve conflicts. but i took your block of code